### PR TITLE
Add a database store for messages

### PR DESCRIPTION
**Pull request by delthas**
Co-authored-by: Simon Ser <contact@emersion.fr>

### CODE REVIEW

**Comment by delthas** [comment]
Just RFC for now.

Being tested.

**Comment by delthas** [comment]
Tested successfully on SQLite and PostgreSQL. Ready for review.

**Comment by delthas** [comment]
Rebased and reverted behavior of `log fs ""` back to an error rather than memory

File: database/database.go [hunk 1]
```diff
@@ -10,8 +10,25 @@ import (
...
 
+type MessageTarget struct {
+	Name          string
+	LatestMessage time.Time
+}
+
+type MessageOptions struct {
+	AfterID  int64
+	After    time.Time
+	Before   time.Time
```
Comment (emersion) [inline database/database.go:24]: Nit: can we rename these to `AfterTime`/`BeforeTime`?

File: database/database.go [hunk 2]
```diff
@@ -40,6 +57,11 @@ type Database interface {
 	ListWebPushSubscriptions(ctx context.Context, userID, networkID int64) ([]WebPushSubscription, error)
 	StoreWebPushSubscription(ctx context.Context, userID, networkID int64, sub *WebPushSubscription) error
 	DeleteWebPushSubscription(ctx context.Context, id int64) error
+
+	GetMessageLastID(ctx context.Context, networkID int64, name string) (int64, error)
+	StoreMessage(ctx context.Context, networkID int64, name string, msg *irc.Message) (int64, error)
+	ListMessageLastPerTarget(ctx context.Context, networkID int64, options MessageOptions) ([]MessageTarget, error)
+	ListMessages(ctx context.Context, networkID int64, name string, options MessageOptions) ([]*irc.Message, error)
```
Comment (emersion) [inline database/database.go:64]: Nit: can we pass the options as pointers here? There are quite a few fields in there.

File: contrib/migrate-logs/main.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	for _, user := range users {
+		log.Printf("Migrating logs for user: %s\n", user.Username)
+
+		networks, err := db.ListNetworks(ctx, user.ID)
+		if err != nil {
+			log.Fatalf("unable to get networks for user: #%d %s", user.ID, user.Username)
+		}
+
+		for _, network := range networks {
+			log.Printf("Migrating logs for network: %s\n", network.Name)
```
Comment (emersion) [inline contrib/migrate-logs/main.go:68]: Can we extract this loop's body into a separate function?

**Review by emersion — COMMENTED** [review]
Looks pretty solid overall! A few comments below, nothing major.

File: Makefile [hunk 4]
```diff
@@ -1,7 +1,7 @@
 GO ?= go
 RM ?= rm
 SCDOC ?= scdoc
-GOFLAGS ?=
+GOFLAGS ?= --tags sqlite_fts5
```
[the remark this answers is not in the document]
Comment (delthas) [reply Makefile]: 🙅

**Comment by emersion** [comment]
It seems like CI is unhappy about this PR.

File: database/postgres.go [hunk 5]
```diff
@@ -173,6 +199,29 @@ var postgresMigrations = []string{
...
 	`ALTER TABLE "User" ADD COLUMN downstream_interacted_at TIMESTAMP WITH TIME ZONE`,
+	`
+		CREATE TABLE MessageTarget (
+			id SERIAL PRIMARY KEY,
+			network INTEGER NOT NULL REFERENCES "Network"(id) ON DELETE CASCADE,
+			target TEXT NOT NULL,
+			UNIQUE(network, target)
+		);
+
+		CREATE TEXT SEARCH DICTIONARY "public.search_simple" (
```
Comment (emersion) [inline database/postgres.go:210]: Seems like this migration is out-of-sync with the schema.

**Review by emersion — APPROVED** [review]
Thanks a lot!

**Pull request merged**